src/rate_limiter.rs

```diff
@@ -0,0 +1,247 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::time::{Duration, Instant};
+
+use crate::TokenBucket;
+
+pub struct RateLimiter<'a, K> {
+    buckets: HashMap<K, TokenBucket<'a>>,
+    clock: &'a (dyn Fn() -> Instant + Sync),
+}
+
+impl<'a, K: Eq + Hash> RateLimiter<'a, K> {
+    pub fn new() -> Self {
+        Self::with_timer(&Instant::now)
+    }
+
+    fn with_timer(clock: &'a (dyn Fn() -> Instant + Sync)) -> Self {
+        Self {
+            buckets: HashMap::new(),
+            clock,
+        }
+    }
+
+    pub fn set_limit(&mut self, key: K, limit: usize, interval: Duration) {
+        self.buckets
+            .insert(key, TokenBucket::with_timer(limit, interval, self.clock));
+    }
+
+    pub fn consume(&self, key: K, tokens: usize) -> bool {
+        self.buckets
+            .get(&key)
+            .map(|bucket| bucket.consume(tokens))
+            .unwrap_or(true)
+    }
+}
+
+impl<'a, K: Eq + Hash> Default for RateLimiter<'a, K> {
```

Comment:
Nit. We could drop `Eq + Hash` if we split the `impl` block above. Otherwise, ignore it :)

Reply:
That's what `cargo clippy` generated for me :)

Reply:
Yeah, that's because `new()` requires it now, but it does not have to.